### PR TITLE
[FIXED JENKINS-21618] Refactoring how statistics are calculated

### PR DESCRIPTION
**Pull request by stephenc**
This is a request for review.

@jglick @kohsuke any thoughts?

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/jenkins/1480)

### CODE REVIEW

**Comment by stephenc** [comment]
@reviewbybees and @kohsuke ping

File: core/src/main/java/hudson/model/LoadStatistics.java [hunk 1]
```diff
@@ -67,15 +69,59 @@
...
+    /**
+     * Number of executors not executing and available to take work and how it changes over time.
+     * @since FIXME
+     */
+    @Exported
+    public final MultiStageTimeSeries availableExecutors;
+
+    /**
      * Number of total executors and how it changes over time.
+     * @deprecated use {@link #onlineExecutors}
```
Comment (kohsuke) [inline core/src/main/java/hudson/model/LoadStatistics.java]: Doc should point out `totalExecutors==onlineExecutors` now for backward compatibility support.

File: core/src/main/java/jenkins/model/UnlabeledLoadStatistics.java [hunk 2]
```diff
@@ -73,4 +82,56 @@ public int computeTotalExecutors() {
...
+
+    private static class UnlabeledNodesIterable implements Iterable<Node> {
+
+        @Override
+        public Iterator<Node> iterator() {
+            return new UnlabeledNodesIterator();
+        }
+    }
+
+    private static class UnlabeledNodesIterator implements Iterator<Node> {
```
Comment (kohsuke) [inline core/src/main/java/jenkins/model/UnlabeledLoadStatistics.java]: Opportunity to use `FilterIterator`.

File: core/src/main/java/hudson/model/LoadStatistics.java [hunk 3]
```diff
@@ -180,17 +249,71 @@ public Api getApi() {
...
+
+    /**
+     * Returns {@code true} is the specified {@link SubTask} from the {@link Queue} should be counted.
+     * @param item the {@link SubTask}
+     * @return {@code true} IFF the specified {@link SubTask} from the {@link Queue} should be counted.
+     * @since FIXME
+     */
+    protected abstract boolean matches(SubTask item);
+
+    public LoadStatisticsSnapshot computeSnapshot() {
```
Comment (kohsuke) [inline core/src/main/java/hudson/model/LoadStatistics.java]: This is a key method exposed to the rest of the world, so it should have javadoc.

**Comment by kohsuke** [comment]
It'd be nice if the commit message contains a rationale and IIUC, a reference to JENKINS-21618.

**Comment by kohsuke** [comment]
Aside from those mostly minor comments, I think the change looks overall sensible.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 4]
```diff
@@ -3492,8 +3492,7 @@ public void run() {
                     LOGGER.severe(String.format("Shutting down VM as requested by %s from %s",
                                                 exitUser, exitAddr));
                     // Wait 'til we have no active executors.
-                    while (isQuietingDown
-                           && (overallLoad.computeTotalExecutors() > overallLoad.computeIdleExecutors())) {
+                    while (isQuietingDown && (overallLoad.computeSnapshot().getBusyExecutors() > 0)) {
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: Did this not conflict with #1610?

**Comment by thomassuckow** [comment]
This conflicts with my PR #1569 and I am wondering how to proceed. The logic for counting "buildable items" (subtasks) is duplicated between the loadstatistics and queue. Both improperly use == and the queue additionally handles the null label improperly.

Should I worry about deduplication of code?
If not, I can fix both locations and move on with my life.
If we should de-duplicate, it opens a can of worms because we would probably need to make the queue snapshot visible to the loadstatistics. We could just make it package scope I suppose.

Thoughts?

File: core/src/main/java/jenkins/model/UnlabeledLoadStatistics.java [hunk 5]
```diff
@@ -73,4 +83,42 @@ public int computeTotalExecutors() {
...
     }
+
+    @Override
+    protected Iterable<Node> getNodes() {
+        return nodes;
+    }
+
+    @Override
+    protected boolean matches(SubTask item) {
+        return true;
```
Comment (thomassuckow) [inline core/src/main/java/jenkins/model/UnlabeledLoadStatistics.java]: Shouldn't the only match tasks with a "null" label?

File: core/src/main/java/hudson/model/Label.java [hunk 6]
```diff
@@ -107,6 +108,16 @@ public int computeTotalExecutors() {
...
             }
+
+            @Override
+            protected Set<Node> getNodes() {
+                return Label.this.getNodes();
+            }
+
+            @Override
+            protected boolean matches(SubTask item) {
+                return item.getAssignedLabel() == Label.this;
```
Comment (thomassuckow) [inline core/src/main/java/hudson/model/Label.java]: This should probably be .equals(). Labels are not guaranteed to be the same instance even if they have the same name.

**Pull request merged**